src/lib/pubkey/kyber/kyber_common/kyber.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Crystals Kyber key encapsulation mechanism
+ * Based on the public domain reference implementation by the
+ * designers (https://github.com/pq-crystals/kyber)
+ *
+ * Further changes
+ * (C) 2021-2022 Jack Lloyd
+ * (C) 2021-2022 Manuel Glaser and Michael Boric, Rohde & Schwarz Cybersecurity
+ * (C) 2021-2022 René Meusel and Hannes Rantzsch, neXenio GmbH
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_KYBER_COMMON_H_
+#define BOTAN_KYBER_COMMON_H_
+
+#include <botan/pk_keys.h>
+#include <botan/exceptn.h>
+#include <botan/der_enc.h>
+
+#include <botan/internal/pk_ops_impl.h>
+
+#if !defined(BOTAN_HAS_KYBER_90S) && !defined(BOTAN_HAS_KYBER)
+   static_assert(false, "botan module 'kyber_common' is useful only when enabling modules 'kyber', 'kyber_90s' or both");
+#endif
```

Comment:
We opted to introduce three modules `kyber` (modern), `kyber_90s` and `kyber_common`. The latter contains the actual algorithm and supporting math code. The former two contain adapters to the required symmetric algorithms for XOF, KDF and more.

Hence, users enabling `kyber_common` exclusively at `./configure.py` wouldn't end up with a functional kyber implementation. They'd need at least one of the symmetric algorithm adapters in the other mentioned modules. This `static_assert` is an attempt to warn them of their mistake at build time.

To the best of our knowledge, Botan's module system doesn't have a notion of an "internal module" for such use cases, right? Is that potentially something we'd want to introduce?

Reply:
Replying to @mouse07410 inline, to keep the context:

> IMHO, it makes no sense to allow kyber_common for explicit configuration.
I'd have only kyber and kyber_90s as "configurable", each of which would force inclusion of kyber_common. Adding it explicitly without one of the other two makes no sense, disabling it when at least one of the others is enabled doesn't make sense either.

I agree with you! Though, given the inner workings of Botan's module system, we didn't figure out how to do that without introducing the `kyber_common` module. I.e. this solution is a technical crutch for the time being. Is there a better way, @randombit?

Reply:
Ironically there *used* to be a good way of doing this, a module could have an `A or B` dependency, which would work perfectly in this situation. It looks like I removed this in the otherwise unrelated commit ba8a26f17d - the functionality hadn't been used in some time at that point, and in that PR I was making other changes to dependency resolution to handle conditional dependencies based on OS features and apparently decided to remove `|` at the same time.

This current setup is certainly going to be confusing to end users. I'll see how hard it would be to add alternatives back, probably not bad. Then we would have `kyber` which would depend on `kyber_modern | kyber_90s`.

Reply:
Sounds good, if it's not too complicated to get alternatives back in. Would the module `kyber` default to the modern mode then? Or would users have to explicity enable `kyber,kyber_modern`?

Note that currently the `kyber` module is the modern mode and `kyber_90s` the 90s mode. So end users would need to explicitly only enable `kyber_common` to get a broken build. So naively using `kyber` would be fine. We hoped this would avoid confusion at least a bit.

Reply:
As an alternative, we might take a look at creating "internal modules" (i.e. by adding some new keyword in `info.txt`) so that the current `kyber_common` module could be "required" by `kyber` and `kyber_90s` but wouldn't show up in the user-facing `./configure.py --list-modules`.